### PR TITLE
Dynamic input file generation for nvme and nvmf devices.

### PR DESCRIPTION
**Pull request by manvanthar**
Dynamic input file generation for nvme and nvmf devices.

Code changes include
1. Config file change to include all the block device entries
2. Input file reorder for nvme, FC and nvmf devices to differentiate input values
3. changes to differentiate FC and nvmf devices.
4. new logic to identify nvmf config and devices
5. logic to differentiate b/w nvmf RHEL8 and other Distros
6. correction of pci_devices values

### CODE REVIEW

**Comment by manvanthar** [comment]
For RHEL8.x


# lspci
002c:01:00.0 Fibre Channel: Emulex Corporation LPe37000/LPe38000 Series 32Gb/64Gb Fibre Channel Adapter (rev 10)
002c:01:00.1 Fibre Channel: Emulex Corporation LPe37000/LPe38000 Series 32Gb/64Gb Fibre Channel Adapter (rev 10)
0040:01:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X
0048:01:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM173X


# python3 ./pci_info.py --show-info
[{'adapter_description': 'Emulex Corporation LPe37000/LPe38000 Series '
                         '32Gb/64Gb Fibre Channel Adapter',
  'adapter_id': '10df:f500:1014:06c2',
  'adapter_type': 'nvmf',
  'class': 'scsi_host',
  'disks': ['/dev/mapper/mpatha',
            '/dev/mapper/mpathb',
            '/dev/mapper/mpathc',
            '/dev/mapper/mpathd',
            '/dev/mapper/mpathe',
            '/dev/mapper/mpathf',
            '/dev/mapper/mpathg',
            '/dev/mapper/mpathh',
            '/dev/mapper/mpathi'],
  'driver': ['nvme_fc', 'nvme_fabrics', 'nvme_core'],
  'firmware': '14.0.574.12',
  'functions': ['002c:01:00.0', '002c:01:00.1'],
  'infiniband_interfaces': [],
  'interfaces': ['host2', 'host3'],
  'is_root_disk': False,
  'mpath_disks': ['/dev/mapper/mpatha',
                  '/dev/mapper/mpathb',
                  '/dev/mapper/mpathc',
                  '/dev/mapper/mpathd',
                  '/dev/mapper/mpathe',
                  '/dev/mapper/mpathf',
                  '/dev/mapper/mpathg',
                  '/dev/mapper/mpathh',
                  '/dev/mapper/mpathi'],
  'mpath_wwids': ['eui.a80000000000032000507608108001b3',
                  'eui.a80000000000032100507608108001b3',
                  'eui.a8000000000002a900507608108001b3',
                  'eui.a8000000000002aa00507608108001b3',
                  'eui.a80000000000032200507608108001b3',
                  'eui.a8000000000002ac00507608108001b3',
                  'eui.a80000000000032300507608108001b3',
                  'eui.a8000000000002ad00507608108001b3',
                  'eui.a8000000000002ab00507608108001b3'],
  'pci_root': '002c:01:00',
  'slot': 'U780C.ND0.WZS001N-P0-C6'},
 {'adapter_description': 'Samsung Electronics Co Ltd NVMe SSD Controller '
                         'PM173X',
  'adapter_id': '144d:a824:1014:065f',
  'adapter_type': 'nvme',
  'class': 'nvme',
  'disks': ['/dev/nvme0n1'],
  'driver': 'nvme',
  'firmware': 'REV.SN49',
  'functions': ['0040:01:00.0'],
  'infiniband_interfaces': [],
  'interfaces': ['nvme0'],
  'is_root_disk': False,
  'mpath_disks': [],
  'mpath_wwids': [],
  'pci_root': '0040:01:00',
  'slot': 'U780C.ND0.WZS001N-P1-C6'},
 {'adapter_description': 'Samsung Electronics Co Ltd NVMe SSD Controller '
                         'PM173X',
  'adapter_id': '144d:a824:1014:065f',
  'adapter_type': 'nvme',
  'class': 'nvme',
  'disks': ['/dev/nvme1n1'],
  'driver': 'nvme',
  'firmware': 'REV.SN49',
  'functions': ['0048:01:00.0'],
  'infiniband_interfaces': [],
  'interfaces': ['nvme1'],
  'is_root_disk': False,
  'mpath_disks': [],
  'mpath_wwids': [],
  'pci_root': '0048:01:00',
  'slot': 'U780C.ND0.WZS001N-P1-C5'}]



# python3 ./pci_info.py --create-config
09:05:24 INFO    : python avocado-setup.py --run-suite host_io_nvmf_rhel8_002c_01_00_fvt,host_io_nvme_0040_01_00_fvt,host_io_nvme_0048_01_00_fvt --input-file io_input.txt


# cat io_input.txt 
[io_nvmf_rhel8_002c_01_00_fvt]
disk = "/dev/mapper/mpatha"
lv_disks = "/dev/mapper/mpathb"
htx_disks = "/dev/mapper/mpathc"
disks = "/dev/mapper/mpathd /dev/mapper/mpathe /dev/mapper/mpathf /dev/mapper/mpathg /dev/mapper/mpathh /dev/mapper/mpathi"
wwids = "eui.a80000000000032000507608108001b3 eui.a80000000000032100507608108001b3 eui.a8000000000002a900507608108001b3 eui.a8000000000002aa00507608108001b3 eui.a80000000000032200507608108001b3 eui.a8000000000002ac00507608108001b3 eui.a80000000000032300507608108001b3 eui.a8000000000002ad00507608108001b3 eui.a8000000000002ab00507608108001b3"
pci_device = "002c:01:00.0"
pci_devices = "002c:01:00.0 002c:01:00.1"
module = "nvme_fc nvme_fabrics nvme_core"
dir = "/mnt"
lv_size = "10G"
count = 30
fsstress_loop = 3
num_of_dlpar = 30
num_of_hotplug = 20
iteration = 30
run_type = "rpm"
manageSystem = ""
userid = ""
password = ""
hmc_username = ""
hmc_pwd = ""
htx_rpm_link = ""
switch_name = ""

[io_nvme_0040_01_00_fvt]
devices = "nvme0"
disk = "/dev/nvme0n1"
lv_disks = "/dev/nvme0n1"
htx_disks = "/dev/nvme0n1"
disks = "/dev/nvme0n1"
pci_device = "0040:01:00.0"
pci_devices = "0040:01:00.0"
module = "nvme"
count = 10
fsstress_loop = 3
num_of_dlpar = 20
num_of_hotplug = 20
iteration = 20
run_type = "rpm"
max_freeze = 6
function = 4
err = 1
namespace = 10
utilization = "100"
only_io = True
manageSystem = ""
hmc_username = 
hmc_pwd = 
htx_rpm_link = ""
firmware_url = ""

[io_nvme_0048_01_00_fvt]
devices = "nvme1"
disk = "/dev/nvme1n1"
lv_disks = "/dev/nvme1n1"
htx_disks = "/dev/nvme1n1"
disks = "/dev/nvme1n1"
pci_device = "0048:01:00.0"
pci_devices = "0048:01:00.0"
module = "nvme"
count = 10
fsstress_loop = 3
num_of_dlpar = 20
num_of_hotplug = 20
iteration = 20
run_type = "rpm"
max_freeze = 6
function = 4
err = 1
namespace = 10
utilization = "100"
only_io = True
manageSystem = ""
hmc_username = 
hmc_pwd = 
htx_rpm_link = ""
firmware_url = ""





For RHEL9.X



# lspci
0136:a0:00.0 Fibre Channel: Emulex Corporation LPe37000/LPe38000 Series 32Gb/64Gb Fibre Channel Adapter (rev 10)
0136:a0:00.1 Fibre Channel: Emulex Corporation LPe37000/LPe38000 Series 32Gb/64Gb Fibre Channel Adapter (rev 10)
0136:a0:00.2 Fibre Channel: Emulex Corporation LPe37000/LPe38000 Series 32Gb/64Gb Fibre Channel Adapter (rev 10)
0136:a0:00.3 Fibre Channel: Emulex Corporation LPe37000/LPe38000 Series 32Gb/64Gb Fibre Channel Adapter (rev 10)


# python3 ./pci_info.py --show-info
/root/tests/./pci_info.py:30: DeprecationWarning: The SafeConfigParser class has been renamed to ConfigParser in Python 3.2. This alias will be removed in future versions. Use ConfigParser directly instead.
  CONFIGFILE = configparser.SafeConfigParser()
[{'adapter_description': 'Emulex Corporation LPe37000/LPe38000 Series '
                         '32Gb/64Gb Fibre Channel Adapter',
  'adapter_id': '10df:f500:1014:06c1',
  'adapter_type': 'nvmf',
  'class': 'scsi_host',
  'disks': ['/dev/nvme2n6',
            '/dev/nvme2n5',
            '/dev/nvme2n4',
            '/dev/nvme2n3',
            '/dev/nvme2n2',
            '/dev/nvme2n1'],
  'driver': ['nvme_fc', 'nvme_fabrics', 'nvme_core'],
  'firmware': '14.2.598.0',
  'functions': ['0136:a0:00.0', '0136:a0:00.1', '0136:a0:00.2', '0136:a0:00.3'],
  'infiniband_interfaces': [],
  'interfaces': ['host3', 'host4', 'host5', 'host6'],
  'is_root_disk': False,
  'mpath_disks': [],
  'mpath_wwids': [],
  'pci_root': '0136:a0:00',
  'slot': None}]


# python3 ./pci_info.py --create-config
/root/tests/./pci_info.py:30: DeprecationWarning: The SafeConfigParser class has been renamed to ConfigParser in Python 3.2. This alias will be removed in future versions. Use ConfigParser directly instead.
  CONFIGFILE = configparser.SafeConfigParser()
09:13:57 INFO    : python avocado-setup.py --run-suite host_io_nvmf_0136_a0_00_fvt --input-file io_input.txt

# cat io_input.txt 
[io_nvmf_0136_a0_00_fvt]
device = "/dev/nvme2n6"
disk = "/dev/nvme2n5"
lv_disks = "/dev/nvme2n4"
htx_disks = "/dev/nvme2n3 /dev/nvme2n2 /dev/nvme2n1"
disks = "/dev/nvme2n3 /dev/nvme2n2 /dev/nvme2n1"
pci_device = "0136:a0:00.0"
pci_devices = "0136:a0:00.0 0136:a0:00.1 0136:a0:00.2 0136:a0:00.3"
module = "nvme_fc nvme_fabrics nvme_core"
dir = "/mnt"
lv_size = "5G"
count = 20
fsstress_loop = 2
num_of_dlpar = 20
iteration = 20
run_type = 'rpm'
max_freeze = 6
function = 4
err = 1
manageSystem = ""
userid = ""
password = ""
hmc_username = ""
hmc_pwd = ""
htx_rpm_link = ""
switch_name = ""

File: config/wrapper/pci_input.conf [hunk 1]
```diff
@@ -1,20 +1,57 @@
 [io_nvme_fvt]
+devices = interfaces:0
```
Comment (abdhaleegit) [inline config/wrapper/pci_input.conf:2]: interfaces for nvme bucket, any test needs an interace?

File: config/wrapper/pci_input.conf [hunk 2]
```diff
@@ -23,6 +60,38 @@ pci_device = functions:0
...
+disk = disks:0
+htx_disks = disks:all
+disks = disks:all
+pci_device = functions:0
+pci_devices = functions:all
+module = driver
+
+[io_lsi3008_fvt]
+disk = disks:0
+htx_disks = disks:all
```
Comment (abdhaleegit) [inline config/wrapper/pci_input.conf:73]: current CR takes different disk for htx too, may be 1 or 2 devices dedicated to htx

File: lib/pci.py [hunk 3]
```diff
@@ -39,6 +41,152 @@ def get_domains():
...
 
+def  is_nvmf():
+    """
+    Verify if nvmf is configured
+
+    :return: True if nvmf is configure, False if nvmf is not configured
+    """
+
+    try:
+        nvme_list_subsys_output_bytes = subprocess.check_output(['nvme', 'list-subsys'])
```
Comment (abdhaleegit) [inline lib/pci.py:52]: you can reuse runcmd from library https://github.com/lop-devops/tests/blob/5f18b85850e53a143dedfdb2a1d29cfdca040e86/lib/helper.py#L31

File: lib/pci.py [hunk 4]
```diff
@@ -39,6 +41,152 @@ def get_domains():
...
+            return True
+        else:
+            return False
+
+    except subprocess.CalledProcessError as e:
+        print(f"Error running 'nvme list-subsys': {e}")
+        return False
+
+def is_rhel8():
+    """
```
Comment (abdhaleegit) [inline lib/pci.py:64]: generic api's will go to helper.py   and pci specific apis will be in pci.py

File: lib/pci.py [hunk 5]
```diff
@@ -39,6 +41,152 @@ def get_domains():
...
+    In the nvme list-subsystem output find the nvme host specific to FC  node names
+
+    :param unique node_names for each FC host
+
+    :return: nvme host names for the corresponding FC node names.
+    """
+    nvme_names = []
+    try:
+        nvme_list_subsys_output = subprocess.check_output(['nvme', 'list-subsys'], universal_newlines=True)
```
Comment (abdhaleegit) [inline lib/pci.py:151]: if suited, make use of runcmd() from helper

File: lib/pci.py [hunk 6]
```diff
@@ -403,23 +540,39 @@ def get_pci_name(pci_address):
         return " ".join(pci_name)
     return ""
 
-
-def get_driver(pci_address):
+def get_driver(adapter_type, pci_address):
```
Comment (abdhaleegit) [inline lib/pci.py:543]: keep adapter_type optional, ie  as second param and the legacy code would work only for pci_address

**Review by abdhaleegit — CHANGES_REQUESTED** [review]

File: config/wrapper/pci_input.conf (hunk 1 above)
Comment (manvanthar) [reply config/wrapper/pci_input.conf:2]: Hi Abdul, Thanks for the review.  This is based on the existing input file which was looking for values such as nvme0, nvme1.   which is provided by the interfaces parameter.

File: config/wrapper/pci_input.conf (hunk 2 above)
Comment (manvanthar) [reply config/wrapper/pci_input.conf:73]: Sure will provide separate disk for HTX

File: config/wrapper/pci_input.conf (hunk 1 above)
Comment (abdhaleegit) [reply config/wrapper/pci_input.conf:2]: https://github.com/lop-devops/tests/blob/47de43e0efad17c1eb23685dd864d42e84d0d288/config/inputs/io_nvme_input.txt#L2

the nvme value is taken as device: and not as interfaces   in input file  ..can you check once again

File: lib/pci.py (hunk 6 above)
Comment (manvanthar) [reply lib/pci.py:543]: get_driver function will handle both Network and Block devices.    nvmf is the  one off case which has multiple drivers to be accounted.

File: lib/pci.py [hunk 7]
```diff
@@ -403,23 +540,39 @@ def get_pci_name(pci_address):
...
+                parts = line.strip().split()
+                if len(parts) > 0:
+                    drivers.append(parts[0])
+
+            return drivers
+
+        except subprocess.CalledProcessError as e:
+            print(f"Error running 'lsmod | grep nvme_fc': {e}")
+            return ""
+
```
Comment (PraveenPenguin) [inline lib/pci.py:568]: why not using runcmd

File: lib/pci.py (hunk 6 above)
Comment (PraveenPenguin) [reply lib/pci.py:543]: yes, agree with Abdul. may have  `adapter_type` has `None` as default

File: lib/pci.py (hunk 4 above)
Comment (PraveenPenguin) [reply lib/pci.py:64]: even there get distro utility in helper lib see if we can enhance same if not can be used directly

File: config/wrapper/pci_input.conf [hunk 8]
```diff
@@ -23,6 +60,38 @@ pci_device = functions:0
...
+pci_devices = functions:all
+module = driver
+
+[io_lsi3961_fvt]
+disk = disks:0
+htx_disks = disks:all
+disks = disks:all
+pci_device = functions:0
+pci_devices = functions:all
+module = driver
```
Comment (PraveenPenguin) [inline config/wrapper/pci_input.conf:85]: is this reflection of bucket has new test case added or deletion

**Review by PraveenPenguin — COMMENTED** [review]
@manvanthar @abdhaleegit sorry for the late review, have reviewed comment on the style as the design  aspect still gathering info

File: lib/pci.py (hunk 4 above)
Comment (manvanthar) [reply lib/pci.py:64]: Done resolved

File: lib/pci.py (hunk 6 above)
Comment (manvanthar) [reply lib/pci.py:543]: As discussed with Abdul offline.  this is not required. marking it as resolved.

File: config/wrapper/pci_input.conf (hunk 8 above)
Comment (manvanthar) [reply config/wrapper/pci_input.conf:85]: Enabling the Dynamic input selection code to handle the new IO technology.

File: config/wrapper/pci_input.conf (hunk 1 above)
Comment (manvanthar) [reply config/wrapper/pci_input.conf:2]: From input file perspective, the terms,
a. device means block device
b. interface means network interface

inside the code,  the term interface is generic,  it captures the data of both block and network device.  
It has been coded in such a way that,  depending on the device type the corresponding value is gathered.

**Comment by PraveenPenguin** [comment]
@manvanthar can you please squash , commits in valid commit(s) as we good to go

**Comment by manvanthar** [comment]
Closing the PR

**Pull request closed without merging**